RationaleMCP/0031/differences.md

```diff
@@ -151,6 +151,84 @@ end 'M';
+Hence, the first `when`-clause triggered by `initial()` in full Modelica needs to be turned into `initial equation` form in Flat Modelica, with `reinit`-equations replaced by equality-equations.
+This also means that in Flat Modelica, the triggering condition `initial()` will have the same effect as the triggering condition `true and initial()`, namely that they will never trigger the `when`-clause because the expression never undergoes a positive edge.
+
+The implicit initial equations `x = pre(x)` in full Modelica (in case no `when`-clause is activated with `initial()`) need to be made explicit in Flat Modelica.
+
+Regarding `when`-equations inside `if`-equations and `for`-equations, full Modelica only allows this where the `if`-equation conditions and `for`-equation ranges are parameter expressions.
+Hence, it is only with a small loss of generality that it is being assumed that these conditions and ranges should be possible to evaluate during translation, allowing an `if`-equation to be reduced to one of its branches, or a `for`-equation to be unrolled.
+
+
+## When-Statements
+
+The `when`-statements in Flat Modelica are more restricted compared to full Modelica.
+In summary:
+- `when`-statements have no meaning at all for the initialization problem:
+  * No special treatment of `initial()` as a `when`-clause trigger expression.
+
+**Example 1**: Consider the following full Modelica `when`-statement with a clause triggered by `initial()`:
+```
+  Real x;
+  Real y;
+algorithm
+  x := 0.5;
+  x := x + time;
+  when {x^2 > 2.0, initial()} then
+    y := x;
+  end when;
+  x := x + 0.5;
+```
+This can be expressed with the help of an `if`-statement in Flat Modelica:
+```
+  Real 'x';
+  Real 'y';
+algorithm
+  'x' := 0.5;
+  'x' := 'x' + time;
+  if initial() then /* Right before or right after when-statement does not matter. */
+    'y' := 'x';
+  end if;
+  when 'x'^2 > 2.0 then
+    'y' := 'x';
+  end when;
+  'x' := 'x' + 0.5;
+```
+
+**Example 2**: The `if`-equation trick has issues when relying on discrete-time variability inside the `when`-clause:
+```
+  Real x;
+  Real y;
+algorithm
+  x := 0.5;
+  x := x + time;
+  when {x^2 > 2.0, initial()} then
+    y := pre(x);
+  end when;
+  x := x + 0.5;
+```
+Note that the following is illegal since the argument of `pre` needs to be discrete-time:
+```
+  if initial() then
+    'y' := pre('x'); /* 'x' is continuous-time, since no longer inside when-clause. */
+  end if;
+```
+Open questions:
+- What are the other ways one might depend on the discrete-time variability inside a `when`-clause?
+- Could there be a resolution in utilizing that `pre('x')` appears guarded by `initial()` in the `if`-statement, as this means that `pre('x')` refers to an independent variable of the initialization problem?
+- Better to simply allow `when`-statements to be activated during initialization by `initial()` in Flat Modelica?
```

Comment:
Commenting in responds to this comment by @HansOlsson:
> At first I thought it didn't matter, but thinking more I see a larger issue - and my conclusion is that either we don't change this, or we have some special construct (not sure if "if initial()" suffices).

Yes, this is very similar to what I concluded with the open questions left here.  Right now, if we don't have ideas to try, I'd say there's not enough time to try to do anything about the `when`-statements.  I think we should just leave a comment that unlike `when`-equations, there is no difference between full Modelica and Flat Modelica when it comes to `when`-statements.

Reply:
Web Meeting (Hans, Henrik, Gerd, Martin, Oliver)

Discussed the when-equation and when-statement examples:
* For when-equations the `when initial()` case can be simplified as proposed.
* For when-statements the discussion confirmed that there is no straight-forward solution.

Conclusion:
Leave the text for when-equations as is.
Replace the text for when-statements with a comment that no differences w.r.t. to full Modelica are applied in flat Modelica. A comment shall explain that the case above is not as easily solved as for when-equations.

Decision:
Fix the PR accordingly [Henrik]
Approve in next meeting.